File: jsatorb/jsatorb-eclipse-service/deprecated/main/java/fr/isae/supaero/nanostar/RestifiedEclipseCalculator.java
package fr.isae.supaero.nanostar;

import org.hipparchus.util.Pair;
import org.json.simple.JSONArray;
import org.orekit.data.DataProvidersManager;
import org.orekit.data.DirectoryCrawler;
import org.orekit.time.AbsoluteDate;

import java.io.File;
import java.text.SimpleDateFormat;
import java.util.List;

import org.json.simple.JSONObject;
import org.json.simple.parser.*;

import static spark.Spark.*;

public class
RestifiedEclipseCalculator {
    public static void main(String[] args) {

        options("/*",
                (request, response) -> {

                    String accessControlRequestHeaders = request
                            .headers("Access-Control-Request-Headers");
                    if (accessControlRequestHeaders != null) {
                        response.header("Access-Control-Allow-Headers",
                                accessControlRequestHeaders);
                    }

                    String accessControlRequestMethod = request
                            .headers("Access-Control-Request-Method");
                    if (accessControlRequestMethod != null) {
                        response.header("Access-Control-Allow-Methods",
                                accessControlRequestMethod);
                    }

                    return "OK";
                });
        before((req, res) -> {
            res.header("Access-Control-Allow-Origin", "*");
            res.header("Access-Control-Allow-Headers", "*");
            res.type("application/json");
        });

        File orekitData = new File("./orekit-data");
        DataProvidersManager manager = DataProvidersManager.getInstance();
        manager.addProvider(new DirectoryCrawler(orekitData));

        get("/hello", (req, res)-> {
            return "hello World";
        });

        post("/propagation/eclipses", (req, res) -> {
            String body = req.body();
            SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
            try{
                JSONObject jsonInput = (JSONObject)new JSONParser().parse(req.body());
                JSONObject header = (JSONObject)jsonInput.get("header");
                JSONObject sat = (JSONObject)jsonInput.get("satellite");

                String stringDate = header.get("timeStart").toString();
                long duration = (long)header.get("duration");
                String type = sat.get("type").toString();
                if(type.contains("keplerian")){
                    double sma = Double.parseDouble((String) sat.get("sma"));
                    if(sma < 6371000){
                        return error("bad sma value");
                    }
                    double ecc = Double.parseDouble((String) sat.get("ecc"));
                    double inc = Double.parseDouble((String)sat.get("inc"));
                    double pa = Double.parseDouble((String)sat.get("pa"));
                    double raan = Double.parseDouble((String)sat.get("raan"));
                    double lv = Double.parseDouble((String) sat.get("meanAnomaly"));
                    EclipseCalculator calculator = new EclipseCalculator(new HAL_SatPos(sma,ecc,inc,pa,raan,lv, "keplerian"), formatter.parse(stringDate), duration);
                    return (eclipseToJSON(calculator.getEclipse()));
                }else if(type.contains("cartesian")){
                    double x=Double.parseDouble((String) sat.get("x"));
                    double y=Double.parseDouble((String) sat.get("y"));
                    double z=Double.parseDouble((String) sat.get("z"));
                    double vx=Double.parseDouble((String) sat.get("vx"));
                    double vy=Double.parseDouble((String) sat.get("vy"));
                    double vz=Double.parseDouble((String) sat.get("vz"));
                    EclipseCalculator calculator = new EclipseCalculator(new HAL_SatPos(x,y,z,vx,vy,vz, "cartesian"), formatter.parse(stringDate), duration);
                    return (eclipseToJSON(calculator.getEclipse()));
                }else{
                    return error("bad type");
                }
            } catch (Exception e){
                return error(e.toString());
            }
        });
    }

    public static String error(String errorName){
        return "{\"error\": \""+ errorName +"\" }";
    }

    public static JSONArray eclipseToJSON(List<Pair<AbsoluteDate, AbsoluteDate>> eclipse){
        JSONArray eclipseArray = new JSONArray();
//        String result = "[";
        for(Pair<AbsoluteDate, AbsoluteDate> el : eclipse){
//            result+= "{\"start\": \""+el.getKey().toString() + "\"," + "\"end\" :\"" + el.getValue() + "\" },";
            JSONObject obj = new JSONObject();

            obj.put("start", el.getKey().toString());
            obj.put("end", el.getValue().toString());
            eclipseArray.add(obj);
        }
//        result = result.substring(0, result.length() - 1);

//        return result + "]";
        return eclipseArray;
    }
}

File: jsatorb/jsatorb-eclipse-service/src/RestifiedEclipseCalculator.py
import bottle
from bottle import request, response
import json

from EclipseCalculator import HAL_SatPos, EclipseCalculator
from datetime import datetime

app = application = bottle.default_app()

# the decorator
def enable_cors(fn):
    def _enable_cors(*args, **kwargs):
        # set CORS headers
        response.headers['Access-Control-Allow-Origin'] = '*'
        response.headers['Access-Control-Allow-Methods'] = 'GET, POST, PUT, OPTIONS'
        response.headers['Access-Control-Allow-Headers'] = 'Origin, Accept, Content-Type, X-Requested-With, X-CSRF-Token'

        if bottle.request.method != 'OPTIONS':
            # actual request; reply with the actual response
            return fn(*args, **kwargs)
    return _enable_cors

@app.route('/propagation/eclipses', method=['OPTIONS','POST'])
@enable_cors
def EclipseCalculatorREST():
    response.content_type = 'application/json'
    
    data = request.json
    print(json.dumps(data))
    
    stringDateFormat = '%Y-%m-%dT%H:%M:%S'

    try:
        header = data['header']
        sat = data['satellite']

        stringDateInit = str( header['timeStart'] )
        stringDateEnd = str( header['timeEnd'] )

        typeSat = str( sat['type'] )
        if 'keplerian' in typeSat:
            sma = float( sat['sma'] )
            if sma < 6371000:
                return ValueError('bad sma value')
            ecc = float( sat['ecc'] )
            inc = float( sat['inc'] )
            pa = float( sat['pa'] )
            raan = float( sat['raan'] )
            lv = float( sat['meanAnomaly'] )
            calculator = EclipseCalculator(HAL_SatPos(sma, ecc, inc, pa, raan, lv, 'keplerian'),
                datetime.strptime(stringDateInit, stringDateFormat), datetime.strptime(stringDateEnd, stringDateFormat))
            return eclipseToJSON( calculator.getEclipse() )

        elif 'cartesian' in typeSat:
            x = float( sat['x'] )
            y = float( sat['y'] )
            z = float( sat['z'] )
            vx = float( sat['vx'] )
            vy = float( sat['vy'] )
            vz = float( sat['vz'] )
            calculator = EclipseCalculator(HAL_SatPos(x, y, z, vx, vy, vz, 'cartesian'), 
                datetime.strptime(stringDateFormat, stringDateFormat), duration)
            return eclipseToJSON( calculator.getEclipse() )

        else:
            return error('bad type')

    except Exception as e:
        return error(type(e).__name__)

def error(errorName):
    return '{"error": "' + errorName + '"}'

def eclipseToJSON(eclipse):
    eclipseDictionary = []

    for el in eclipse:
        obj = {}
        obj['start'] = el[0].toString()
        obj['end'] = el[1].toString()
        eclipseDictionary.append(obj)

    return json.dumps(eclipseDictionary)
        

if __name__ == '__main__':
    bottle.run(host = '127.0.0.1', port = 4567)